### PR TITLE
Use phpunit groups to run database code in isolation

### PR DESCRIPTION
**Pull request by DeepDiver1975**
There have already been some code pieces to look into:
- [ ] Why do unit tests for preview generation require database access? (Yes - filesystem foo - but could be structured differently)
- [ ] Pure tests on the server container - why? 
- [ ] DIContainerTest
- [ ] avatar tests
- [x] avatar controller tests #20510
- [ ] stream wrapper tests
- [ ] OCA\Provisioning_API\Tests\AppsTest needs mocks
- [ ] ShareControllerTest needs to mock file system
- [ ] All LDAP unit tests require DB access - even those who already mock IConfig etc :fish: -y
- [ ] Test_UrlGenerator - basically the UrlGenerator needs to inject the router object
- [ ] https://github.com/owncloud/core/issues/20771
- [ ] Test_Util_CheckServer on postgresql

TODO:
- review all unit test in group DB - is real database access really required here?
- review all still failing unit tests - is real database access required?
- think about more separation of existing classes to get more isolated testing (e.g. preview generation)

### CODE REVIEW

**Comment by rullzer** [comment]
>(Yes - filesystem foo - but could be structured differently)

Seems like a good time to revive https://github.com/owncloud/3rdparty/pull/185

**Comment by MorrisJobke** [comment]
I also thought about this some time ago. I didn't know of this grouping. :)

File: autotest.sh [hunk 1]
```diff
@@ -260,14 +260,27 @@ function execute_tests {
 	if [[ "$_XDEBUG_CONFIG" ]]; then
 		export XDEBUG_CONFIG=$_XDEBUG_CONFIG
 	fi
+	if [ -z "$TEST_SELECTION" ]; then
```
Comment (MorrisJobke) [inline autotest.sh]: Why is this needed? The variable is not used below.

Comment (DeepDiver1975) [inline autotest.sh]: :see_no_evil: :question: :question:

Comment (MorrisJobke) [inline autotest.sh]: There is no need to change the value from empty to `all` because below only `DB` and `NODB` are checked. Why adding code that is useless and only confuses ;)

Comment (DeepDiver1975) [inline autotest.sh]: > only confuses ;)

oh - it's that easy to confuse you :smiling_imp:

**Comment by DeepDiver1975** [comment]
@nickvergessen any clue why this in only on php7? THX

```
03:50:59 1) OC\Settings\Controller\EncryptionControllerTest::testStartMigrationSuccessful
03:50:59 OCP\AutoloadNotAllowedException: Autoload path not allowed: /ssd/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/apps/encryption/lib/migration.php
03:50:59 
03:50:59 /ssd/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/lib/autoloader.php:154
03:50:59 /ssd/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/lib/autoloader.php:174
03:50:59 /ssd/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/tests/settings/controller/EncryptionControllerTest.php:94

**Comment by nickvergessen** [comment]
Well encryption app is not enabled by default, so it's not enabled, so the paths are not allowed to be loaded.

Why this only happens in php7? I think because 
```php
		$migration = $this->getMockBuilder('\\OCA\\Encryption\\Migration')
			->disableOriginalConstructor()->getMock();
```
uses Reflection class and that behaviour might have changed, due to type hinting.
So the fix is to add the apps/encryption manually to the loaded paths:
https://github.com/owncloud/core/blob/master/tests/bootstrap.php#L11

**Comment by DeepDiver1975** [comment]
> Well encryption app is not enabled by default, so it's not enabled, so the paths are not allowed to be loaded.

but we enable it in https://github.com/owncloud/core/blob/master/tests/enable_all.php

**Comment by DeepDiver1975** [comment]
> but we enable it in https://github.com/owncloud/core/blob/master/tests/enable_all.php

https://github.com/owncloud/core/blob/master/autotest.sh#L259

**Comment by DeepDiver1975** [comment]
Let's get this in @MorrisJobke @PVince81 @icewind1991 @blizzz THX

File: tests/lib/util.php [hunk 2]
```diff
@@ -167,55 +167,6 @@ function testGetInstanceIdGeneratesValidId() {
 	}
 
 	/**
-	 * Tests that the home storage is not wrapped when no quota exists.
```
Comment (nickvergessen) [inline tests/lib/util.php]: huu? Is this intended?

File: tests/lib/testcase.php [hunk 3]
```diff
@@ -49,6 +51,16 @@ protected function getTestTraits() {
 	}
 
 	protected function setUp() {
+		// detect database access
+		if (!$this->IsDatabaseAccessAllowed()) {
```
Comment (nickvergessen) [inline tests/lib/testcase.php]: This will break all apps I think, not sure how to do it better,
so please drop a mail to the mailing list asap?

Comment (DeepDiver1975) [inline tests/lib/testcase.php]: > so please drop a mail to the mailing list asap?

Or we disable this for apps running on travis - we need this for core only from my pov

File: tests/lib/util.php (hunk 2 above)
Comment (DeepDiver1975) [inline tests/lib/util.php]: hmmmm - rebase-fu - let me double check

File: tests/lib/testcase.php (hunk 3 above)
Comment (nickvergessen) [inline tests/lib/testcase.php]: well it can help in apps as well, only some tests require the DB and then they can make use of it.

About disabling for apps, the tests should pass locally as well, so the only way I could imagine is checking the class namespace: if start with OCA and not with OCA\Dav|Files|...\ then ignore

File: tests/lib/util.php (hunk 2 above)
Comment (DeepDiver1975) [inline tests/lib/util.php]: Ah .. I moved them to a file of it's own: https://github.com/owncloud/core/commit/cb74b575afbf9a086e2eb4980ce45641e65a8bef#diff-17da1c724f4cabe27a25e2108e779f65R29

File: tests/lib/testcase.php (hunk 3 above)
Comment (DeepDiver1975) [inline tests/lib/testcase.php]: > well it can help in apps as well, only some tests require the DB and then they can make use of it.

sure - but I'd take care of this in a PR of it's own - let's just make sure this PR does not break apps.

**Comment by blizzz** [comment]
> All LDAP unit tests require DB access - even those who already mock IConfig etc :fish: -y

We're dealing with more tables than just those dealt with IConfig.

**Comment by nickvergessen** [comment]
:+1:

File: tests/lib/util.php (hunk 2 above)
Comment (MorrisJobke) [inline tests/lib/util.php]: But this commit is missing in this PR.

Comment (nickvergessen) [inline tests/lib/util.php]: he just rebased afterwards again

Comment (nickvergessen) [inline tests/lib/util.php]: https://github.com/owncloud/core/pull/20248/files#diff-17da1c724f4cabe27a25e2108e779f65

**Comment by MorrisJobke** [comment]
:+1:

**Comment by oparoz** [comment]
> sure - but I'd take care of this in a PR of it's own - let's just make sure this PR does not break apps.

#fail. Breaks Gallery tests. There should have been a mailing list announcement to let devs know what they need to do to solve the new error message they may be getting.

**Comment by nickvergessen** [comment]
Well we didn't break Travis, but yes should have sent an email nevertheless

**Comment by oparoz** [comment]
Sure, Travis still works, but it's not a long term solution to bypass the check for one CI system only. People still test locally and might be using other CI solutions.

**Comment by DeepDiver1975** [comment]
> #fail. Breaks Gallery tests

can you please link the error? THX

**Comment by nickvergessen** [comment]
@oparoz please see https://mailman.owncloud.org/pipermail/devel/2015-December/001810.html

**Comment by oparoz** [comment]
@DeepDiver1975 - I can't link to it, you simply see the integration tests fail in the console because they need the DB, unlike unit tests.

@nickvergessen - Thanks for that. I should be able to modify the profile for those tests and get them to pass again.
